### PR TITLE
feat(anta): Added test case to verify Spanning Tree (RPVST) state(stable topology)

### LINKED ISSUE #782: Add Test case to verify Spanning Tree (RPVST) state
Test Criteria:
Ignore anything under the Topology= NoStp section. For all other Topologies returned, number of changes should be less than `ExpectedNoOfChange` indicating a stable topology

Sample output:

DLF#show spanning-tree topology status detail
Topology: NoStp
  Mapped Vlans: 1501-1506
  Cpu:                      forwarding (1 changes, last 179 days, 6:58:55 ago)
  Ethernet49/1:             forwarding (1 changes, last 179 days, 6:58:42 ago)
**Topology: Vl1500**
  Mapped Vlans: 1500
  Cpu:                      forwarding (**1 changes**, last 179 days, 6:58:55 ago)
  Port-Channel100:          forwarding (**1 changes**, last 179 days, 6:58:55 ago)
  Vxlan1:                   forwarding (**1 changes**, last 179 days, 6:58:55 ago)
 
**Topology: Vl1999**
  Mapped Vlans: 1999
  Cpu:                      forwarding (**1 changes**, last 179 days, 6:58:42 ago)
  Port-Channel100:          forwarding (**1 changes**, last 179 days, 6:58:42 ago)
  Vxlan1:                   forwarding (**1 changes**, last 179 days, 6:58:42 ago)

### PR DESCRIPTION
**Pull request by vitthalmagadum**
# Description

Verifies the number of changes across all interfaces in the Spanning Tree Protocol (STP) topology is below a threshold.

    Expected Results
    ----------------
    * Success: The test will pass if the total number of changes across all interfaces is less than the specified threshold.
    * Failure: The test will fail if the total number of changes across all interfaces meets or exceeds the specified threshold,
    indicating potential instability in the topology.

Fixes #782 

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have run pre-commit for code linting and typing (`pre-commit run`)
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes (`tox -e testenv`)

### CODE REVIEW

File: anta/tests/stp.py [hunk 1]
```diff
@@ -259,3 +259,63 @@ def test(self) -> None:
...
+            self.result.is_failure("No STP topology configured")
+            return
+
+        # verifies all available topologies except the "NoStp" topology.
+        stp_topologies.pop("NoStp", None)
+
+        # Verifies the number of changes across all interfaces
+        for topology, topology_details in stp_topologies.items():
+            interfaces = {
+                interface: {"numChanges": num_of_changes}
```
Comment (MaheshGSLAB) [inline anta/tests/stp.py:309]:
```suggestion
                interface: {"Number of changes": num_of_changes}
```

File: anta/tests/stp.py [hunk 2]
```diff
@@ -259,3 +259,63 @@ def test(self) -> None:
...
+                for interface, details in topology_details.get("interfaces", {}).items()
+                if (num_of_changes := details.get("numChanges")) > self.inputs.threshold
+            }
+            if interfaces:
+                failures["topologies"][topology] = interfaces
+
+        if failures["topologies"]:
+            self.result.is_failure(
+                f"The following Spanning Tree Protocol (STP) topology(s) are not configured or number of changes not within the threshold:{failures}"
```
Comment (MaheshGSLAB) [inline anta/tests/stp.py:318]: do we have the case of not configured STP topology?

File: tests/units/anta_tests/test_stp.py [hunk 3]
```diff
@@ -324,4 +324,63 @@
...
+                    "Cist": {
+                        "interfaces": {
+                            "Cpu": {"state": "forwarding", "numChanges": 1, "lastChange": 1723990624.735365},
+                            "Port-Channel5": {"state": "forwarding", "numChanges": 1, "lastChange": 1723990624.7353542},
+                        }
+                    },
+                    "NoStp": {
+                        "interfaces": {
+                            "Cpu": {"state": "forwarding", "numChanges": 1, "lastChange": 1723990624.735365},
+                            "Ethernet1": {"state": "forwarding", "numChanges": 15, "lastChange": 1723990624.7353542},
```
Comment (MaheshGSLAB) [inline tests/units/anta_tests/test_stp.py:343]: how this got pass when changes are 15 which is more then the default 10

File: tests/units/anta_tests/test_stp.py [hunk 4]
```diff
@@ -324,4 +324,63 @@
...
+                        }
+                    },
+                },
+            },
+        ],
+        "inputs": {},
+        "expected": {"result": "success"},
+    },
+    {
+        "name": "failure",
```
Comment (MaheshGSLAB) [inline tests/units/anta_tests/test_stp.py:353]: please add relevant name for failures

File: tests/units/anta_tests/test_stp.py (hunk 3 above)
Comment (vitthalmagadum) [reply tests/units/anta_tests/test_stp.py:343]: verifies all available topologies except the "NoStp" topology.

File: anta/tests/stp.py (hunk 2 above)
Comment (vitthalmagadum) [reply anta/tests/stp.py:318]: Even if STP is enabled globally, specific VLANs might not have STP configured, resulting in no active STP topology.

File: tests/units/anta_tests/test_stp.py (hunk 4 above)
Comment (vitthalmagadum) [reply tests/units/anta_tests/test_stp.py]: Addressed

**Review by MaheshGSLAB — APPROVED** [review]
LGTM

File: anta/tests/stp.py [hunk 5]
```diff
@@ -259,3 +259,63 @@ def test(self) -> None:
...
+    name = "VerifyStpTopologyChanges"
+    description = "Verifies the number of changes across all interfaces in the Spanning Tree Protocol (STP) topology is below a threshold."
+    categories: ClassVar[list[str]] = ["stp"]
+    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [AntaCommand(command="show spanning-tree topology status detail", revision=1)]
+
+    class Input(AntaTest.Input):
+        """Input model for the VerifyStpTopologyChanges test."""
+
+        threshold: int = 10
+        """The threshold number of changes in the STP topology. Defaults to 10."""
```
Comment (carl-baillargeon) [inline anta/tests/stp.py:291]: Let's remove the default and let the user decides.

File: anta/tests/stp.py [hunk 6]
```diff
@@ -259,3 +259,63 @@ def test(self) -> None:
...
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """Main test function for VerifyStpTopologyChanges."""
+        failures: dict[str, Any] = {"topologies": {}}
+
+        command_output = self.instance_commands[0].json_output
+        if not (stp_topologies := command_output.get("topologies")):
```
Comment (carl-baillargeon) [inline anta/tests/stp.py:299]: Is there a scenario where `topologies` would be an empty dict?

File: anta/tests/stp.py (hunk 5 above)
Comment (vitthalmagadum) [reply anta/tests/stp.py]: Updated.

File: tests/units/anta_tests/test_stp.py [hunk 7]
```diff
@@ -324,4 +324,63 @@
...
+                "The following Spanning Tree Protocol (STP) topology(s) are not configured or number of changes"
+                " not within the threshold:{'topologies': {'Cist': {'Cpu': {'Number of changes': 15}, 'Port-Channel5': {'Number of changes': 15}}}}"
+            ],
+        },
+    },
+    {
+        "name": "failure-topologies-not-configured",
+        "test": VerifyStpTopologyChanges,
+        "eos_data": [
+            {"unmappedVlans": [], "topologies": {}},
```
[the remark this answers is not in the document]
Comment (vitthalmagadum) [reply tests/units/anta_tests/test_stp.py]: No, NoStp topologies will be there. Updated the unit tests.

File: anta/tests/stp.py [hunk 8]
```diff
@@ -259,3 +259,66 @@ def test(self) -> None:
...
+
+        command_output = self.instance_commands[0].json_output
+        stp_topologies = command_output.get("topologies", {})
+
+        # verifies all available topologies except the "NoStp" topology.
+        stp_topologies.pop("NoStp", None)
+
+        # Verify the STP topology(s).
+        if not stp_topologies:
+            self.result.is_failure("None of STP topology is configured.")
```
Comment (carl-baillargeon) [inline anta/tests/stp.py:306]:
```suggestion
            self.result.is_failure("STP is not configured.")
```

File: anta/tests/stp.py [hunk 9]
```diff
@@ -259,3 +259,66 @@ def test(self) -> None:
...
+                interface: {"Number of changes": num_of_changes}
+                for interface, details in topology_details.get("interfaces", {}).items()
+                if (num_of_changes := details.get("numChanges")) > self.inputs.threshold
+            }
+            if interfaces:
+                failures["topologies"][topology] = interfaces
+
+        if failures["topologies"]:
+            self.result.is_failure(
+                f"The following Spanning Tree Protocol (STP) topology(s) are not configured or number of changes not within the threshold:{failures}"
```
Comment (carl-baillargeon) [inline anta/tests/stp.py:321]:
```suggestion
                f"The following Spanning Tree Protocol (STP) topologies are not configured or number of changes not within the threshold:{failures}"
```

File: tests/units/anta_tests/test_stp.py [hunk 10]
```diff
@@ -324,4 +324,98 @@
...
+                        }
+                    },
+                },
+            },
+        ],
+        "inputs": {"threshold": 10},
+        "expected": {"result": "success"},
+    },
+    {
+        "name": "success-rstp",
```
Comment (carl-baillargeon) [inline tests/units/anta_tests/test_stp.py:353]: Can you also add a test with `rapid-pvst` please?

Comment (vitthalmagadum) [reply tests/units/anta_tests/test_stp.py:353]: Added unit test for rapid-pvst.
Thanks!!

File: anta/tests/stp.py (hunk 9 above)
Comment (vitthalmagadum) [reply anta/tests/stp.py]: Updated failure message.

File: anta/tests/stp.py (hunk 8 above)
Comment (vitthalmagadum) [reply anta/tests/stp.py]: Updated. Thanks!!

**Review by carl-baillargeon — APPROVED** [review]
LGTM

**Pull request merged**